Implement sql.Scanner and driver.Value interfaces

bruth opened the pull request:
This adds methods to work seamlessly with SQL statements.

/cc @peterbourgon @tsenart

bruth commented:
CI build failed due to a Coveralls error:

```
$ $HOME/gopath/bin/goveralls -coverprofile=cov.out -service=travis-ci -repotoken $COVERALLS_TOKEN
Bad response status from coveralls: 422
{"message":"Couldn't find a repository matching this job.","error":true}
The command "$HOME/gopath/bin/goveralls -coverprofile=cov.out -service=travis-ci -repotoken $COVERALLS_TOKEN" exited with 1.
```

peterbourgon commented:
We tried this before: https://github.com/oklog/ulid/pull/10

@tsenart Has anything changed in your opinion, or does this PR bring new stuff to the table?

bruth commented:
Sorry I missed! I guess my response to that discussion is that the package could provide a default method so it "just works", but obviously the user can choose if they want to go with it or wrap to serialize as a string instead.

tsenart commented:
> @tsenart Has anything changed in your opinion, or does this PR bring new stuff to the table?

Yeah, my thoughts on this have changed.

> Sorry I missed! I guess my response to that discussion is that the package could provide a default method so it "just works", but obviously the user can choose if they want to go with it or wrap to serialize as a string instead.

I'm cool with that approach as long as those tradeoffs are clearly documented.

tsenart commented:
So, for that matter, providing a default `Scan` implementation that also handles strings would be good. Then those who want to serialize strings only need to implement one method, not two, on their wrapper type.

bruth commented:
Updated to handle scanning strings as well. Assuming this is good, I will quick add a couple tests.

tsenart commented on ulid.go:
```diff
@@ -340,3 +341,24 @@ func (id *ULID) SetEntropy(e []byte) error {
+		return id.UnmarshalBinary(x)
+	}
+
+	return errors.New("ulid: source value must be bytes")
```
With string support, this error message must be adapted.

tsenart commented on ulid.go:
```diff
@@ -340,3 +341,24 @@ func (id *ULID) SetEntropy(e []byte) error {
+	switch x := src.(type) {
+	case nil:
+		return nil
+
```
Please remove the vertical whitespace between the switch cases.

tsenart commented on ulid.go:
```diff
@@ -340,3 +341,24 @@ func (id *ULID) SetEntropy(e []byte) error {
+	return errors.New("ulid: source value must be bytes")
+}
+
+// Value implements the sql/driver.Valuer interface.
```
Let's add clear documentation in this comment that the generated value is of the []byte type and how to do it with strings via the wrapper type approach.

tsenart commented:
Yep, +1 for tests.

tsenart commented:
> Addressed the comments. For Value, I noted that a wrapper type can be created that calls MarshalText. I suppose a code example could be added or even a function that handles this automatically could do it, something like:

I'd just provide a code example in GoDoc.

tsenart commented on ulid_test.go:
```diff
@@ -390,6 +390,32 @@ func TestOverflowHandling(t *testing.T) {
 	}
 }
 
+func TestScanBytes(t *testing.T) {
```
These two tests are mostly identical. Could we merge them into a single table driven test? Something like:

```go
id := ulid.MustNew(123, crand.Reader)

for _, tc := range []struct{
    name string
    in interface{}
    out ULID
    err error
} {
    {"string", id.String(), id, nil},
    {"bytes", id[:], id, nil},
    {"nil", nil, ULID(nil), nil},
    {"other", 44, nil, ErrScanValue},
} {
    tc := tc
    t.Run(tc.name, func(t *testing.T) {
        t.Parallel()

        var out ulid.ULID
        err := out.Scan(tc.in)
        if got, want := out, tc.out; !got.Equal(want) {
            t.Errorf("got ULID %s, want %s", got, want)
        }

         if got, want := fmt.Sprint(err), fmt.Sprint(tc.err); got != want {
             t.Errorf("got err %q, want %q", got, want)
         }
    })
    
}
```

bruth commented:
Good call with the tests. Updated. I also added the example in GoDoc. This is what is looks like:

<img width="834" alt="screen shot 2018-07-12 at 7 26 25 am" src="https://user-images.githubusercontent.com/515142/42630568-300d0ea4-85a5-11e8-86af-0b5572f4b910.png">

tsenart commented:
Ran tests locally. They pass. I'll fix CI today.

imaramos commented:
Just a question, I am storing this in postgres as bytea, I have tried to do encoding in order to be more readable to a human but I am not being able to do it. Shouldn't at least to be able to see the numeric value that the byte array represents?
Thank you

imaramos commented:
I am storing the ulid byte array , as a bytea column in postgres. So I have the byte array stored. But when doing the query it isnt human readable ( because it is only bytes). I am asking if there is a way of show at db level this column as a more human readable value, allowing also do the match with the ULID representation. ( I think I will not be able to do that because  the encoding used is Crockford's base32)

tsenart commented:
If you store bytes in Postgres, you’ll get bytes back when querying. If you need the database to show the string representation of the ULIDs, you must either store that instead or define a Postgres function that can do the encoding itself at query time. For the letter, I found this on Google: https://github.com/schinckel/ulid-postgres

tsenart commented:
I’d recommend you simply store the string representation unless you have space constraints true justify the gains of storing bytes.

imaramos commented:
I have to try that , but it is still in progress so...  About the storing the string, is not about the space but also the efficience and also I will loose the ability to sort right?

peterbourgon commented:
ULID strings are sortable. Whether or not it's efficient enough for your application is a question that can only be answered with representative usage data and profiling.

imaramos commented:
Right, thank you for your help guys!
